src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -1070,46 +1127,131 @@ DatabaseShardImp::findShardIndexToAdd(
     for (int i = 0; i < 40; ++i)
     {
         auto const shardIndex {rand_int(earliestShardIndex(), maxShardIndex)};
-        if (complete_.find(shardIndex) == complete_.end() &&
-            (!incomplete_ || incomplete_->index() != shardIndex) &&
-            preShards_.find(shardIndex) == preShards_.end())
-        {
+        if (shards_.find(shardIndex) == shards_.end())
             return shardIndex;
-        }
     }
 
     assert(false);
     return boost::none;
 }
 
 void
-DatabaseShardImp::setFileStats(std::lock_guard<std::mutex>&)
+DatabaseShardImp::finalizeShard(
+    ShardInfo& shardInfo,
+    bool writeSQLite,
+    std::lock_guard<std::mutex>&)
 {
-    fileSz_ = 0;
-    fdRequired_ = 0;
-    if (!complete_.empty())
+    assert(shardInfo.shard);
+    assert(shardInfo.shard->index() != acquireIndex_);
+    assert(shardInfo.shard->isBackendComplete());
+    assert(shardInfo.state != ShardInfo::State::finalize);
+
+    auto const shardIndex {shardInfo.shard->index()};
+
+    shardInfo.state = ShardInfo::State::finalize;
+    taskQueue_->addTask([this, shardIndex, writeSQLite]() mutable
     {
-        for (auto const& e : complete_)
+        std::shared_ptr<Shard> shard;
         {
-            fileSz_ += e.second->fileSize();
-            fdRequired_ += e.second->fdRequired();
+            std::lock_guard lock(mutex_);
```

Comment:
Is it necessary to acquire this lock to make sure the preshard is still present? If not you could just capture the `shared_ptr` in this lambda, avoiding the additional mutex lock and container search. Since the shard's state gets set to `finalize`, it's not possible for the shard to be removed by a call to `removePreShard`. The only other places it could have been removed are during initialization and destruction. But let me know if I'm mistaken!

Reply:
At this time, the lock and search may be unnecessary but the design will handle upcoming changes. For instance the task dealing with the new shard index selection algorithm will need to drop finalized shards when rotating and we want to avoid finalizing a shard that has been dropped. So going forward, the design assumes simultaneous modification and I believe it will be easier and safer to make that assumption with new features.

Note that `FinalizeShard`  is used to import `preshards` and peer network acquired shards.

Reply:
Sounds good